File: Sorting/Merge Sort/mergeSort.py
numbers = [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0,223,332,223,332]

def merge_sort(array):
    if len(array) == 1:
        return array

    # Split the array into left and right halves
    mid = len(array) // 2
    left = array[:mid]
    right = array[mid:]

    left = merge_sort(left)
    right = merge_sort(right)

    return merge(left, right)

def merge(left, right):
    merged = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] <= right[right_index]:
            merged.append(left[left_index])
            left_index += 1
        else:
            merged.append(right[right_index])
            right_index += 1

    # Append any remaining elements from left and right
    merged.extend(left[left_index:])
    merged.extend(right[right_index:])

    return merged

answer = merge_sort(numbers)
print(answer)

File: Sorting/Merge Sort/mergeSort.ts
// const numbers: number[] = [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0];

// function mergeSort(array: number[]): number[] {
//   if (array.length === 1) {
//     return array;
//   }
//   // Split Array into left and right
//   let left: number[] = [];
//   let right: number[] = [];
//   for (let i = 0; i < array.length; i++) {
//     if (i % 2 === 0) {
//       left.push(array[i]);
//     } else {
//       right.push(array[i]);
//     }
//   }
//   return merge(mergeSort(left), mergeSort(right));
// }

// function merge(left: number[], right: number[]): number[] {
//   let merged: number[] = [];
//   while (left.length > 0 && right.length > 0) {
//     if (left[0] <= right[0]) {
//       merged.push(left.shift() as number);
//     } else {
//       if (left[0] > right[0]) {
//         merged.push(right.shift() as number);
//       }
//     }
//   }
//   while (left.length > 0) {
//     merged.push(left.shift() as number);
//   }
//   while (right.length > 0) {
//     merged.push(right.shift() as number);
//   }
//   return merged;
// }

// const answer: number[] = mergeSort(numbers);
// console.log(answer);
